src/txdb.cpp

```diff
@@ -264,11 +280,39 @@ bool CCoinsViewDB::BatchWrite(CCoinsMap &mapCoins,
     
     for (CCswNullifiersMap::iterator it = cswNullifies.begin(); it != cswNullifies.end();) {
         const std::pair<uint256, libzendoomc::ScFieldElement>& position = it->first;
-        BatchWriteCswNullifier(batch, position.first, position.second, it->second.entered);
+        BatchWriteCswNullifier(batch, position.first, position.second, (it->second.flags == CCswNullifiersCacheEntry::FRESH));
         CCswNullifiersMap::iterator itOld = it++;
         cswNullifies.erase(itOld);
     }
 
+    // Store cert Data Hashes
+    for (CCertDataHashMap::iterator it = certDataHashes.begin(); it != certDataHashes.end();) {
+        const std::pair<uint256, int>& position = it->first;
+        uint256 scId = position.first;
+        int epochId = position.second;
+        std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement> prevData;
+        libzendoomc::ScFieldElement cumulativeCertData;
+
+        switch(it->second.flags) {
+            case CCertDataHashCacheEntry::Flags::FRESH:
+            case CCertDataHashCacheEntry::Flags::DIRTY:
+                if (GetCertData(scId, epochId - 1, prevData)) {
+                    cumulativeCertData = calculateCumulativeCertDataHash(prevData.first, prevData.second);
+                }
+                BatchWriteCertDataHash(batch, scId, epochId, it->second.certDataHash, cumulativeCertData);
```

Comment:
I have a doubt that we do `cumulativeCertData` computation here (which also requires `GetCertData` read operation during the `BatchWrite`. As for me, this is a part of business logic that we need to put to cache method `AddCertHashData`. So, the `cumulativeCertData` should be a part of `CCertDataHashCacheEntry`.

Moreover, we don't need to recalculate `CCertDataHashCacheEntry` in case if there is no epoch switch compare to the last cert data stored.